tree/dataframe/src/RDFInterfaceUtils.cxx

```diff
@@ -777,10 +777,14 @@ BookVariationJit(const std::vector<std::string> &colNames, std::string_view vari
    const auto funcName = DeclareFunction(parsedExpr.fExpr, parsedExpr.fVarNames, exprVarTypes);
    const auto type = RetTypeOfFunc(funcName);
 
-   if (type.rfind("ROOT::VecOps::RVec", 0) != 0)
+   if (type.rfind("ROOT::VecOps::RVec", 0) != 0) {
+      // Avoid leak
+      delete upcastNodeOnHeap;
```

Comment:
This cures the symptoms and maybe fine.  However the caller is not informed of this deletion (the pointer to the `shared_ptr` is copied).  Assumingly the callers is not catching the exception and so this should still work.
However it also indicates that the calling code is not exception safe (neither is this fixed version) and may or may not leak other resources.  (This change is only improving the case where *this* function throw the exception), it will still have the leak if any other function it calls throw an exception.

Reply:
I agree with all of your points. This is I believe a very faulty behaviour in the RDataFrame JIT machinery. Unfortunately, it has been there for years, so I am yet to undergo a full refactoring of the RDF JIT. If you believe this ,albeit small, improvement is not enough, I will go for that refactoring.

Reply:
Just to give more context to what I mean, here is a commit from 2018 where the `MakeSharedOnHeap` function (i.e. one of the functions of the JIT machinery that causes the fault memory management) not only was already present but it had a comment saying
```
this function is meant to only be used by TInterface's action methods, and should be deprecated as soon as we find a better way to make jitting work
```

https://github.com/root-project/root/commit/a43693f718d04ea473099d2ec15bc0b518a71df2

Whereas currently that same function lost that comment (now mostly moved to the sibling `MakeWeakOnHeap` but stripped of the `deprecate asap` part)

https://github.com/root-project/root/commit/cdebbee475edb75705d2090ab1fd158aafde4743

Reply:
I am still confused but I see that the pointer is not used __after__ the call to `BookVariationJit` so the change is indeed 'okay'.

However, it seems that the pointer is always leaking in the success case, is that right? (that may or may not be in addition to the leak of the pointee).

Reply:
The pointer is always leaking, that is the intended behaviour. The memory is then managed in the `JitVariationHelper` function [here](https://github.com/root-project/root/blob/50a1bddacc1cbd22d7ce0e184db42bc04e9b0088/tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx#L577). The call to this function is what is being constructed in the lines after the throw of the `std::runtime_error`, where the `varyInvocation` string is being populated and will be JITted afterwards

Reply:
If the pointer is always leaking we need-to/can:
(a) merge this as is, it is no worse that the current situation.
(b) work on a long term fix that does not require leaking memory (even if it is only the 16 bits of the shared pointer themselves)

[leaking the shared_pointer also means the underlying memory is never deleted, isn't it ... or is there a force delete somewhere]

Reply:
The deletion happens at the end of JIT execution,  [here](https://github.com/root-project/root/blob/50a1bddacc1cbd22d7ce0e184db42bc04e9b0088/tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx#L577), [here](https://github.com/root-project/root/blob/50a1bddacc1cbd22d7ce0e184db42bc04e9b0088/tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx#L621), [here](https://github.com/root-project/root/blob/50a1bddacc1cbd22d7ce0e184db42bc04e9b0088/tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx#L530) and [here](https://github.com/root-project/root/blob/50a1bddacc1cbd22d7ce0e184db42bc04e9b0088/tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx#L462-L464)